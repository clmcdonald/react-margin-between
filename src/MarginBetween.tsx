import React, { CSSProperties, PropsWithChildren } from 'react';

interface StyleProp {
  style?: CSSProperties;
}

interface MarginBetweenProps {
  marginHorizontal?: string;
  marginVertical?: string;
}

const MarginBetween: React.FC<MarginBetweenProps> = (props) => {
  const getChildrenWithMargins = (): React.ReactNode => {
    return React.Children.map(props.children, (child, i) => {
      const item = child as React.ReactElement<
        React.PropsWithChildren<StyleProp>
      >;
      const itemStyle: CSSProperties = item.props.style || {};
      const childProps: Partial<PropsWithChildren<StyleProp>> = {};
      childProps.style = itemStyle;
      if (i < React.Children.count(props.children) - 1) {
        if (props.marginHorizontal)
          childProps.style.marginRight = props.marginHorizontal;
        if (props.marginVertical)
          childProps.style.marginBottom = props.marginVertical;
      }
      if (i > 0) {
        if (props.marginHorizontal)
          childProps.style.marginLeft = props.marginHorizontal;
        if (props.marginVertical)
          childProps.style.marginTop = props.marginVertical;
      }

      return React.cloneElement(item, childProps);
    });
  };

  return <React.Fragment>{getChildrenWithMargins()}</React.Fragment>;
};

export default MarginBetween;
