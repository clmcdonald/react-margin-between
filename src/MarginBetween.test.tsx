import React from 'react';
import MarginBetween from './MarginBetween';
import { render } from '@testing-library/react';

describe('MarginBetween tests', () => {
  describe('no margins', () => {
    test('matches snapshot if no children are passed', () => {
      const { asFragment } = render(<MarginBetween />);

      expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot if one child passed', () => {
      const { asFragment } = render(
        <MarginBetween>
          <div>child1</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot if two children passed', () => {
      const { asFragment } = render(
        <MarginBetween>
          <div>child1</div>
          <div>child2</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('horizontal margin', () => {
    test('matches snapshot if one child passed', () => {
      const { asFragment } = render(
        <MarginBetween marginHorizontal='1em'>
          <div>child1</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot if two children passed', () => {
      const { asFragment } = render(
        <MarginBetween marginHorizontal='1em'>
          <div>child1</div>
          <div>child2</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('vertical margin', () => {
    test('matches snapshot if one child passed', () => {
      const { asFragment } = render(
        <MarginBetween marginVertical='1em'>
          <div>child1</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot if two children passed', () => {
      const { asFragment } = render(
        <MarginBetween marginVertical='1em'>
          <div>child1</div>
          <div>child2</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('both margins', () => {
    test('matches snapshot if one child passed', () => {
      const { asFragment } = render(
        <MarginBetween marginVertical='1em' marginHorizontal='2em'>
          <div>child1</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot if two children passed', () => {
      const { asFragment } = render(
        <MarginBetween marginVertical='1em' marginHorizontal='2em'>
          <div>child1</div>
          <div>child2</div>
        </MarginBetween>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
