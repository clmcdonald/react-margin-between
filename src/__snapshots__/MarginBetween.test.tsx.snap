// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarginBetween tests both margins matches snapshot if one child passed 1`] = `
<DocumentFragment>
  <div>
    child1
  </div>
</DocumentFragment>
`;

exports[`MarginBetween tests both margins matches snapshot if two children passed 1`] = `
<DocumentFragment>
  <div
    style="margin-right: 2em; margin-bottom: 1em;"
  >
    child1
  </div>
  <div
    style="margin-left: 2em; margin-top: 1em;"
  >
    child2
  </div>
</DocumentFragment>
`;

exports[`MarginBetween tests horizontal margin matches snapshot if one child passed 1`] = `
<DocumentFragment>
  <div>
    child1
  </div>
</DocumentFragment>
`;

exports[`MarginBetween tests horizontal margin matches snapshot if two children passed 1`] = `
<DocumentFragment>
  <div
    style="margin-right: 1em;"
  >
    child1
  </div>
  <div
    style="margin-left: 1em;"
  >
    child2
  </div>
</DocumentFragment>
`;

exports[`MarginBetween tests no margins matches snapshot if no children are passed 1`] = `<DocumentFragment />`;

exports[`MarginBetween tests no margins matches snapshot if one child passed 1`] = `
<DocumentFragment>
  <div>
    child1
  </div>
</DocumentFragment>
`;

exports[`MarginBetween tests no margins matches snapshot if two children passed 1`] = `
<DocumentFragment>
  <div>
    child1
  </div>
  <div>
    child2
  </div>
</DocumentFragment>
`;

exports[`MarginBetween tests vertical margin matches snapshot if one child passed 1`] = `
<DocumentFragment>
  <div>
    child1
  </div>
</DocumentFragment>
`;

exports[`MarginBetween tests vertical margin matches snapshot if two children passed 1`] = `
<DocumentFragment>
  <div
    style="margin-bottom: 1em;"
  >
    child1
  </div>
  <div
    style="margin-top: 1em;"
  >
    child2
  </div>
</DocumentFragment>
`;
